docs/vm-configuration.md

```diff
@@ -0,0 +1,265 @@
+libvirtd daemon has a privilege level which is equivalent to that of libvirtd
+itself, which is root. Thus applications calling virDomainDefineXML (and many
+other APIs) which accept configuration input from untrusted users, to feed into
+an XML document, must be very careful in their data handling.
+
+
+Config states
+-------------
+
+Naively one might expect a virtual machine to have a single document describing
+its configuration. In practice, this is unreasonably restrictive.
+
+The configuration provided by a tenant or management application will typically
+not express every single frontend or backend setting. When the XML document is
+provided to libvirt, it will fill in any important missing pieces of information
+in an attempt to fully represent the virtual hardware configuration and backend
+integration. For example, if the tenant did not supply PCI addresses, libvirt
+will assign them prior to starting QEMU. Even when libvirt fills in missing
+data this is only ever libvirt's best attempt at a complete virtual hardware
+configuration. For example, for ISA serial ports, libvirt does not record the
+hardware I/O port associated with the serial port, since that is not currently
+a configurable option & thus can be assumed stable without being explicitly
+represented. A future libvirt or QEMU, may allow I/O port configuration, and
+thus when upgrading to newer libvirt, the XML document previously considered
+to be complete may have yet more defaults filled in.
+
+Applications that wish to be able to start the guest multiple times, while
+guaranteeing a stable hardware ABI, need to capture the augmented configuration
+and use it on subsequent boot attempts. Depending on which libvirt APIs are
+used, the augmented configuration can be obtained without starting the guest,
+or at the same time as starting a guest.
+
+When a guest is started, some aspects of backend configuration may be
+dynamically filled in. For example, an automatically chosen VNC port will be
+reported in the XML document. This is information that an application will
+require for interaction with the guest for as long as it is running, but once
+shutoff the information can be discarded.
+
+Running guests can have a variety of configuration changes made to them, in
+particular, hotplugging & unplugging of virtual devices such as disks, NICs,
+etc. These operations are not all guaranteed to be synchronous, nor are they
+even guaranteed to complete at all. For example, unplugging a PCI device
+requires co-operation from the guest OS. A guest that is still running the
+BIOS, or early boot up, may never honour the unplug request. Malicious or buggy
+guests can ignore a request forever too, or take an unexpectedly long time to
+complete it.  There is thus a distinction between the configuration that a
+running guests currently has, vs what it is desired to have, which is only
+resolved when the unplug operation completes.
```

Comment:
It is impossible to tell whether a guest completed a hotplug request or not. From QEMU's POV, it has made the PCI device available, and issued the ACPI trigger to notify the guest. Whether the guest does anything with that is invisible, but it also does not matter from host POV.

On hotunplug, however, things are critical - QEMU issues the ACPI trigger to notify the guest, and the guest must respond before QEMU can complete the unplug operation. This is visible because libvirt won't emit any event until that completion happens.